### PR TITLE
Overlays: rework private options

### PR DESCRIPTION
**Pull request by alexanderPolosatov**
Meaningful changes:
1. Popover and Tooltip default position value is not string anymore, it's an object
2. `_position` cache is calculated in the single function. It's used everywhere instead of direct option value getting
3. Elements which scroll triggers overlay hiding are `wrapper` element parents now. So position does not affect `closeOnTargetScroll` option behaviour anymore

### CODE REVIEW

File: js/ui/overlay/ui.overlay.js [hunk 1]
```diff
@@ -502,12 +472,44 @@ const Overlay = Widget.inherit({
 
     _normalizePosition: function() {
         const position = this.option('position');
-        this._position = typeof position === 'function' ? position() : position;
+        const defaultPositionOptions = {
+            of: this.option('target'),
+            boundaryOffset: DEFAULT_BOUNDARY_OFFSET
+        };
+        if(isFunction(position)) {
+            this._position = extend(true, {}, defaultPositionOptions, position());
```
Comment (ksercs) [inline js/ui/overlay/ui.overlay.js]: function passed to `position` property can return alias string. Seems like this case is broken.

Let's use `getOptionValue`?

File: js/ui/overlay/ui.overlay.js [hunk 2]
```diff
@@ -502,12 +472,44 @@ const Overlay = Widget.inherit({
 
     _normalizePosition: function() {
         const position = this.option('position');
```
Comment (ksercs) [inline js/ui/overlay/ui.overlay.js]: const { position, target } = this.option();

File: js/ui/overlay/ui.overlay.js [hunk 3]
```diff
@@ -502,12 +472,44 @@ const Overlay = Widget.inherit({
...
+            }
+        }
     },
 
     _getAnimationConfig: function() {
         let animation = this.option('animation');
         if(isFunction(animation)) animation = animation.call(this);
+        const target = this.option('target');
+
+        if(isDefined(target)) {
```
Comment (ksercs) [inline js/ui/overlay/ui.overlay.js]: oh, I don't think it's a good idea to move `_initTarget` function code to this function.

`_initTarget` was called only in `initOptions` and `target` option changed handler.
But this function is called every time overlay shows or hides

File: js/ui/overlay/ui.overlay.js [hunk 4]
```diff
@@ -564,7 +566,6 @@ const Overlay = Widget.inherit({
 
         this._normalizePosition();
```
Comment (ksercs) [inline js/ui/overlay/ui.overlay.js:586]: do we really need this call now?

File: js/ui/overlay/ui.overlay.js [hunk 5]
```diff
@@ -849,27 +850,18 @@ const Overlay = Widget.inherit({
...
 
-        if(devices.real().deviceType === 'desktop') {
-            $parents = $parents.add(window);
-        }
-
-        this._proxiedTargetParentsScrollHandler = this._proxiedTargetParentsScrollHandler
-            || (e => { this._targetParentsScrollHandler(e); });
-
-        eventsEngine.off($().add(this._$prevTargetParents), scrollEvent, this._proxiedTargetParentsScrollHandler);
+        eventsEngine.off($().add(this._$prevTargetParents), scrollEvent, this._scrollHandler);
```
Comment (ksercs) [inline js/ui/overlay/ui.overlay.js]: why we cannot get rid of `$().add`. It'll be a problem during renovation

File: js/ui/overlay/ui.overlay.js [hunk 6]
```diff
@@ -290,11 +289,11 @@ const Overlay = Widget.inherit({
 
         this._toggleViewPortSubscription(true);
         this._initHideTopOverlayHandler(this.option('hideTopOverlayHandler'));
+        this._scrollHandler = (e => { this._targetParentsScrollHandler(e); });
```
Comment (ksercs) [inline js/ui/overlay/ui.overlay.js]: Can we name this variable more specific?

Comment (ksercs) [inline js/ui/overlay/ui.overlay.js]: If we use it only inside of `_toggleParentsScrollSubscription` shouldn't we cache both handler and "parents" there as single object?

Something like
```
this._parentScrollSubsciptionInfo= {
    prevTargets: ...,
    targets: ...,
    handler: ...
}

```

File: js/ui/overlay/ui.overlay.js [hunk 7]
```diff
@@ -1292,7 +1284,7 @@ const Overlay = Widget.inherit({
         const container = this.option('container');
         let positionOf = null;
 
-        if(!container && position) {
+        if(!container) {
```
Comment (ksercs) [inline js/ui/overlay/ui.overlay.js]: Why do we change it?

File: js/ui/overlay/ui.overlay.js [hunk 8]
```diff
@@ -1416,7 +1405,7 @@ const Overlay = Widget.inherit({
         this._updateZIndexStackPosition(false);
         this._toggleTabTerminator(false);
 
-        this._actions = null;
+        this._actions = this._scrollHandler = this._$prevTargetParents = null;
```
Comment (ksercs) [inline js/ui/overlay/ui.overlay.js]: it's better to use `undefined` to not show this cache variables in instance object

File: js/ui/overlay/ui.overlay.js [hunk 9]
```diff
@@ -1321,28 +1313,25 @@ const Overlay = Widget.inherit({
...
-            const position = this._transformStringPosition(this._position, POSITION_ALIASES);
-            const resultPosition = positionUtils.setup(this._$content, position);
-
-            return resultPosition;
+            return positionUtils.setup(this._$content, position);
         }
     },
 
-    _transformStringPosition: function(position, positionAliases) {
+    _getPositionOptions: function(positionAliases) {
```
Comment (ksercs) [inline js/ui/overlay/ui.overlay.js]: mb `_getPositionOptionAsObject` or smth like this?

File: testing/tests/DevExpress.ui.widgets.editors/validationMessage.tests.js [hunk 10]
```diff
@@ -33,14 +33,19 @@ QUnit.module('options', moduleSetup, () => {
...
+        const onShowingSpy = sinon.spy();
 
         try {
             $element.appendTo('#qunit-fixture');
+            this._validationMessage.on('showing', onShowingSpy);
 
             this._validationMessage.option('target', $element);
 
-            assert.strictEqual(this._validationMessage.option('position').of, $element, 'position is recalculated');
+            assert.strictEqual(onShowingSpy.callCount, 1, 'position is recalculated');
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/validationMessage.tests.js]: Why did u change this test? I think its initial meaning is lost

File: testing/tests/DevExpress.ui.widgets/dropDownMenu.tests.js [hunk 11]
```diff
@@ -482,12 +482,17 @@ QUnit.module('position', {
         $(element.dxDropDownMenu('instance')._button.$element()).trigger('dxclick');
 
         positionConfig = instance._popup.option('position');
-        assert.equal(positionConfig, instance.option('popupPosition'));
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/dropDownMenu.tests.js:485]: cannot we use `deepEqual` here?

**Review by ksercs — COMMENTED** [review]
Do we need `normalizePosition` call in `renderGeometryImpl` now? Seems like it should be called only on init and appropriate options change

File: testing/tests/DevExpress.ui.widgets.editors/validationMessage.tests.js (hunk 10 above)
Comment (alexanderPolosatov) [reply testing/tests/DevExpress.ui.widgets.editors/validationMessage.tests.js]: because at present, options does not mobified by 'target'. target sets to _position field.
onShowingSpy.callCount is equal to one, because I subscribed on 'showing' event after initial show.

File: js/ui/overlay/ui.overlay.js (hunk 7 above)
Comment (alexanderPolosatov) [reply js/ui/overlay/ui.overlay.js]: because _position already normalized on _init and has 'of' field

File: js/ui/overlay/ui.overlay.js (hunk 5 above)
Comment (alexanderPolosatov) [reply js/ui/overlay/ui.overlay.js]: fixed

File: js/ui/overlay/ui.overlay.js (hunk 4 above)
Comment (alexanderPolosatov) [reply js/ui/overlay/ui.overlay.js:586]: fixed

File: js/ui/overlay/ui.overlay.js (hunk 3 above)
Comment (alexanderPolosatov) [reply js/ui/overlay/ui.overlay.js]: reworkerd

File: js/ui/overlay/ui.overlay.js [hunk 12]
```diff
@@ -878,28 +883,20 @@ const Overlay = Widget.inherit({
...
+        eventsEngine.off(parentsScrollSubscriptionInfo.prevTargets, scrollEvent, parentsScrollSubscriptionInfo.handler);
 
-        eventsEngine.off($().add(this._$prevTargetParents), scrollEvent, this._proxiedTargetParentsScrollHandler);
-
-        if(subscribe && closeOnScroll) {
-            eventsEngine.on($parents, scrollEvent, this._proxiedTargetParentsScrollHandler);
-            this._$prevTargetParents = $parents;
+        const closeOnScroll = this.option('closeOnTargetScroll');
+        if(needSubscribe && closeOnScroll) {
+            let $parents = getElement(this._position.of).parents();
```
Comment (ksercs) [inline js/ui/overlay/ui.overlay.js]: why do we get parents from `position.of`,  not `container`? `position.of` affect visual position but not a markup

File: testing/tests/DevExpress.ui.widgets.editors/validationMessage.tests.js (hunk 10 above)
Comment (ksercs) [reply testing/tests/DevExpress.ui.widgets.editors/validationMessage.tests.js]: Why does `showing` event raise guarantee us position is recalculated? Seems like this test should be written another way. For example, check visual position using `getBoundingClientRect`

File: js/ui/overlay/ui.overlay.js (hunk 1 above)
Comment (alexanderPolosatov) [reply js/ui/overlay/ui.overlay.js]: reworked

File: js/ui/overlay/ui.overlay.js (hunk 12 above)
Comment (alexanderPolosatov) [reply js/ui/overlay/ui.overlay.js]: reworked. at present, we get parents of wrapper element

File: testing/tests/DevExpress.ui.widgets.editors/validationMessage.tests.js (hunk 10 above)
Comment (alexanderPolosatov) [reply testing/tests/DevExpress.ui.widgets.editors/validationMessage.tests.js]: rewirked

File: testing/tests/DevExpress.ui.widgets/overlay.tests.js [hunk 13]
```diff
@@ -2032,26 +2050,10 @@ testModule('close on target scroll', moduleConfig, () => {
         assert.strictEqual(overlay.option('visible'), true, 'overlay should not be hidden if show animation is not completed');
     });
 
-    test('overlay should be hidden if any of jQuery Event target\'s parents were scrolled', function(assert) {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/overlay.tests.js:2035]: Why do we delete this test?

File: testing/tests/DevExpress.ui.widgets/overlay.tests.js [hunk 14]
```diff
@@ -2154,44 +2159,21 @@ testModule('close on target scroll', moduleConfig, () => {
...
-        const targetParent = $target.parent().get(0);
+        const containerParent = $container.parent().get(0);
         $target.detach();
         overlay.hide();
-        const parentEvents = $._data(targetParent).events || {};
+        const parentEvents = $._data(containerParent).events || {};
         assert.strictEqual('scroll' in parentEvents, false, 'scroll unsubscribed');
     });
 
-    test('all opened overlays should be closed on scroll', function(assert) {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/overlay.tests.js:2164]: Why do we delete this test?

File: testing/tests/DevExpress.ui.widgets/overlay.tests.js [hunk 15]
```diff
@@ -2562,22 +2544,68 @@ testModule('target', moduleConfig, () => {
...
+            left: 100
+        });
+
+        const overlay = new Overlay($('#overlay'), {
+            target: $target
         });
+
+        overlay.show();
+
+        const $wrapper = overlay.$content().closest(toSelector(OVERLAY_WRAPPER_CLASS));
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/overlay.tests.js]:
```suggestion
        const $wrapper = overlay.$wrapper();
```

File: js/ui/overlay/ui.overlay.js [hunk 16]
```diff
@@ -1441,7 +1437,7 @@ const Overlay = Widget.inherit({
         this._updateZIndexStackPosition(false);
         this._toggleTabTerminator(false);
 
-        this._actions = null;
+        this._actions = this._scrollHandler = this._$prevTargetParents = null;
```
Comment (ksercs) [inline js/ui/overlay/ui.overlay.js]:
```suggestion
        this._actions = null;
        this._parentsScrollSubscriptionInfo = null;
```

File: js/ui/overlay/ui.overlay.js [hunk 17]
```diff
@@ -276,11 +274,15 @@ const Overlay = Widget.inherit({
 
         this._toggleViewPortSubscription(true);
         this._initHideTopOverlayHandler(this.option('hideTopOverlayHandler'));
+        this._parentsScrollSubscriptionInfo = {
+            handler: (e => { this._targetParentsScrollHandler(e); })
```
Comment (ksercs) [inline js/ui/overlay/ui.overlay.js]: do we really need this brackets`(...)`?

**Review by ksercs — DISMISSED** [review]

File: js/ui/overlay/ui.overlay.js (hunk 17 above)
Comment (alexanderPolosatov) [reply js/ui/overlay/ui.overlay.js]: can be removed

File: testing/tests/DevExpress.ui.widgets/overlay.tests.js (hunk 13 above)
Comment (alexanderPolosatov) [reply testing/tests/DevExpress.ui.widgets/overlay.tests.js:2035]: I think, this test looks like 'funcional', not looks like 'utin-test'. My goal in the next: 'do not have tests on no existing code'. Our exists code is cover by other unit-tests and his test looks like 'extra'.

File: testing/tests/DevExpress.ui.widgets/overlay.tests.js (hunk 14 above)
Comment (alexanderPolosatov) [reply testing/tests/DevExpress.ui.widgets/overlay.tests.js:2164]: see comment about https://github.com/DevExpress/DevExtreme/pull/18385/files#r682836440

Comment (ksercs) [reply testing/tests/DevExpress.ui.widgets/overlay.tests.js:2164]: I returned it back

File: testing/tests/DevExpress.ui.widgets/overlay.tests.js (hunk 13 above)
Comment (ksercs) [reply testing/tests/DevExpress.ui.widgets/overlay.tests.js:2035]: I returned it back

**Review by ksercs — APPROVED** [review]

**Review by DokaRus — APPROVED** [review]

**Review by Zedwag — APPROVED** [review]
